### PR TITLE
Optimize VoxelGrid Filter

### PR DESCRIPTION
**Pull request by facontidavide**
VoxelGrid spends a lot of time sorting.

Spread sort implemented in boost seems to be faster than std::sort.

I get an average speed improvement of 20%, but it would be nice if anyone can reproduce this.

### CODE REVIEW

**Comment by Morwenn** [comment]
As long as you're sorting integers it should indeed always be faster than `std::sort`. I've got a benchmark where several sorting algorithms are used to sort an `std::vector<int>` with 10^7 elements:



`spread_sorter` in the image above is Boost `integer_sort` and as you can see it's consistently faster than `std::sort` except for the *Alternating* pattern.

**Comment by facontidavide** [comment]
@Morwenn . thanks for adding some context.  

You may find a reproducible benchmark here: 

https://github.com/facontidavide/PCL_VoxelGrid_enhanced

image

**Comment by kunaltyagi** [comment]
How about making the sort algorithm a policy or input of the filter? Since other places in PCL could have bad defaults too, and this can be a start in giving the users more freedom.

PCL should still provide good defaults (ala this PR), but we can let the users choose an algorithm that works best for them.

**Comment by facontidavide** [comment]
Honestly, if no one cared to optimize this so far, no one will care to customize it in the future.

The general the user doesn't need to know that VoxelGrid filter uses a sorting algorithm and even less to know about which is the best sorting algorithm for his/her use case.

And if std::sort is always slower, why should I have the "freedom" to select it?

File: filters/include/pcl/filters/impl/voxel_grid.hpp [hunk 1]
```diff
@@ -202,13 +203,18 @@ pcl::getMinMax3D (const typename pcl::PointCloud<PointT>::ConstPtr &cloud,
 
 struct cloud_point_index_idx 
 {
-  unsigned int idx;
+  int idx;
```
Comment (taketwo) [inline filters/include/pcl/filters/impl/voxel_grid.hpp:206]: Could you please explain why this change is needed? Especially considering that line 342 nevertheless instantiates these structures with `unsigned int` cast.

Comment (facontidavide) [reply filters/include/pcl/filters/impl/voxel_grid.hpp:206]: I was following the example in the boost::sort library, but I made a test and this change is not needed in fact.

I should probably revert it to make it more clear

File: filters/include/pcl/filters/impl/voxel_grid.hpp [hunk 2]
```diff
@@ -205,10 +206,15 @@ struct cloud_point_index_idx
   unsigned int idx;
   unsigned int cloud_point_index;
 
-  cloud_point_index_idx (unsigned int idx_, unsigned int cloud_point_index_) : idx (idx_), cloud_point_index (cloud_point_index_) {}
+  cloud_point_index_idx(){}
+  cloud_point_index_idx (int idx_, unsigned int cloud_point_index_) : idx (idx_), cloud_point_index (cloud_point_index_) {}
```
Comment (taketwo) [inline filters/include/pcl/filters/impl/voxel_grid.hpp:210]: Please also revert the change here.

File: filters/include/pcl/filters/impl/voxel_grid.hpp [hunk 3]
```diff
@@ -205,10 +206,15 @@ struct cloud_point_index_idx
   unsigned int idx;
   unsigned int cloud_point_index;
 
-  cloud_point_index_idx (unsigned int idx_, unsigned int cloud_point_index_) : idx (idx_), cloud_point_index (cloud_point_index_) {}
+  cloud_point_index_idx(){}
```
Comment (taketwo) [inline filters/include/pcl/filters/impl/voxel_grid.hpp:209]: Do we need this?

File: filters/include/pcl/filters/impl/voxel_grid.hpp [hunk 4]
```diff
@@ -339,7 +345,9 @@ pcl::VoxelGrid<PointT>::applyFilter (PointCloud &output)
 
   // Second pass: sort the index_vector vector using value representing target cell as index
   // in effect all points belonging to the same output cell will be next to each other
-  std::sort (index_vector.begin (), index_vector.end (), std::less<cloud_point_index_idx> ());
+  //std::sort (index_vector.begin (), index_vector.end (), std::less<cloud_point_index_idx> ());
+  boost::sort::spreadsort::integer_sort(index_vector.begin (), index_vector.end (),
+                                        cloud_point_index_rightshift (), std::less<cloud_point_index_idx> ());
```
Comment (taketwo) [inline filters/include/pcl/filters/impl/voxel_grid.hpp:350]: Is it possible to provide a lambda here?

File: filters/include/pcl/filters/impl/voxel_grid.hpp [hunk 5]
```diff
@@ -339,7 +345,9 @@ pcl::VoxelGrid<PointT>::applyFilter (PointCloud &output)
 
   // Second pass: sort the index_vector vector using value representing target cell as index
   // in effect all points belonging to the same output cell will be next to each other
-  std::sort (index_vector.begin (), index_vector.end (), std::less<cloud_point_index_idx> ());
+  //std::sort (index_vector.begin (), index_vector.end (), std::less<cloud_point_index_idx> ());
```
Comment (taketwo) [inline filters/include/pcl/filters/impl/voxel_grid.hpp:348]: Please remove commented out lines.

File: filters/include/pcl/filters/impl/voxel_grid.hpp (hunk 2 above)
Comment (facontidavide) [reply filters/include/pcl/filters/impl/voxel_grid.hpp:210]: ok

File: filters/include/pcl/filters/impl/voxel_grid.hpp (hunk 3 above)
Comment (facontidavide) [reply filters/include/pcl/filters/impl/voxel_grid.hpp:209]: yes, we do.

File: filters/include/pcl/filters/impl/voxel_grid.hpp (hunk 4 above)
Comment (facontidavide) [reply filters/include/pcl/filters/impl/voxel_grid.hpp:350]: That would be nicer. I will check that out

File: filters/include/pcl/filters/impl/voxel_grid.hpp (hunk 5 above)
Comment (facontidavide) [reply filters/include/pcl/filters/impl/voxel_grid.hpp:348]: ok

File: filters/src/voxel_grid.cpp [hunk 6]
```diff
@@ -377,7 +378,9 @@ pcl::VoxelGrid<pcl::PCLPointCloud2>::applyFilter (PCLPointCloud2 &output)
 
   // Second pass: sort the index_vector vector using value representing target cell as index
   // in effect all points belonging to the same output cell will be next to each other
-  std::sort (index_vector.begin (), index_vector.end (), std::less<cloud_point_index_idx> ());
+  //std::sort (index_vector.begin (), index_vector.end (), std::less<cloud_point_index_idx> ());
```
Comment (kunaltyagi) [inline filters/src/voxel_grid.cpp:381]:
```suggestion
```

**Comment by awesomebytes** [comment]
@facontidavide I had the chance to test the efficacy of your PR (I did need to backport it to PCL 1.8.1 and build from source) and my testing with real data from Velodyne pointclouds gave 23-24% speed improvement, with 0.8% of the computations taking equal or longer than with the previous implementation (could just be my CPU having hickups). I computed it by doing doing an average over comparing pointclouds processed one-to-one and also over total times running some relatively big custom datasets.

**Comment by facontidavide** [comment]
Good to know the actual numbers 😊
